rocBLAS no longer builds

When building rocBLAS from 2.7 release(ie from [1ac2271e72d21fa745c7af87cd5fd102f90cba45](https://github.com/ROCmSoftwarePlatform/rocBLAS/commit/1ac2271e72d21fa745c7af87cd5fd102f90cba45)) we get an error from cmake:

```
Please run this script after at least one build of rocBLAS.
library/src/CMakeFiles/rocblas.dir/build.make:697: recipe for target 'library/src/librocblas.so.0.1' failed
```

Rocblas is not building for ROCM 2.7 even if using install.sh. Please, try Rocm 2.7 .



The rocBLAS checked out from GitHub branch master is failing to build, this is the output:-

 

[ 11%] Building CXX object library/src/CMakeFiles/Tensile.dir/__/__/Tensile/Tools.cpp.o

clang-9: warning: -amdgpu-target argument 'gfx908' is not recognized; using gfx803 instead [-Winvalid-command-line-argument]

clang-9: warning: -amdgpu-target argument 'gfx908' is not recognized; using gfx803 instead [-Winvalid-command-line-argument]

clang-9: warning: -amdgpu-target argument 'gfx908' is not recognized; using gfx803 instead [-Winvalid-command-line-argument]

clang-9: warning: -amdgpu-target argument 'gfx908' is not recognized; using gfx803 instead [-Winvalid-command-line-argument]

clang-9: warning: -amdgpu-target argument 'gfx908' is not recognized; using gfx803 instead [-Winvalid-command-line-argument]

[ 11%] Built target rocblas-test-data

clang-9: error: unable to execute command: Killed

clang-9: error: clang frontend command failed due to signal (use -v to see invocation)

HCC clang version 9.0.0 (/data/jenkins_workspace/compute-rocm-rel-2.7/external/hcc-tot/clang 5c2570257a8bbd74eff632fbc60692ef61ef8ecb) (/data/jenkins_workspace/compute-rocm-rel-2.7/external/hcc-tot/compiler 9a9477021e6998100ff64d1360dcfe64f65cebe5) (based on HCC 2.7.19315-346267d-5c25702-9a94770 )

Target: x86_64-unknown-linux-gnu

Thread model: posix

InstalledDir: /opt/rocm/bin

clang-9: note: diagnostic msg: PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace, preprocessed source, and associated run script.

clang-9: note: diagnostic msg: Error generating preprocessed source(s).

library/src/CMakeFiles/Tensile.dir/build.make:62: recipe for target 'library/src/CMakeFiles/Tensile.dir/__/__/Tensile/Kernels.cpp.o' failed

make[2]: *** [library/src/CMakeFiles/Tensile.dir/__/__/Tensile/Kernels.cpp.o] Error 254

CMakeFiles/Makefile2:110: recipe for target 'library/src/CMakeFiles/Tensile.dir/all' failed

make[1]: *** [library/src/CMakeFiles/Tensile.dir/all] Error 2

Makefile:151: recipe for target 'all' failed

 

xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx

This is my env !

 

apt show rocm-libs -a

 

Version: 2.7.22

Priority: optional

Section: devel

Maintainer: Advanced Micro Devices Inc.

Installed-Size: 13.3 kB

Depends: rocfft, rocrand, rocblas, hipblas, rocsparse, hipsparse, rocalution

Homepage: https://github.com/RadeonOpenCompute/ROCm

Download-Size: 784 B

APT-Sources: http://repo.radeon.com/rocm/apt/debian xenial/main amd64 Packages

Description: Radeon Open Compute (ROCm) Runtime software stack

 

diego@diego-MS-7B09:[master]~/Git/rocBLAS$  dpkg -l | grep rock

ii  rock-dkms                  2.7-22             all                rock-dkms driver in DKMS format.

diego@diego-MS-7B09:[master]~/Git/rocBLAS$ dpkg -l | grep rocm

ii  rocblas                    2.4.0.1471-rocm-re amd64              Radeon Open Compute BLAS library

ii  rocm-clang-ocl             0.4.0.35-rocm-rel- amd64              OpenCL compilation with clang compiler.

ii  rocm-cmake                 0.3.0.93-rocm-rel- amd64              rocm-cmake built using CMake

ii  rocm-dev                   2.7.22             amd64              Radeon Open Compute (ROCm) Runtime software stack

ii  rocm-device-libs           0.0.1              amd64              Radeon Open Compute - device libraries

ii  rocm-dkms                  2.7.22             amd64              Radeon Open Compute (ROCm) Runtime software stack

ii  rocm-opencl                1.2.0-2019080925   amd64              OpenCL/ROCm

ii  rocm-opencl-dev            1.2.0-2019080925   amd64              OpenCL/ROCm

ii  rocm-smi                   1.0.0-167-ge14d23e amd64              System Management Interface for ROCm

ii  rocm-utils                 2.7.22             amd64              Radeon Open Compute (ROCm) Runtime software stack

ii  rocminfo                   1.0.0              amd64              Radeon Open Compute (ROCm) Runtime rocminfo tool

diego@diego-MS-7B09:[master]~/Git/rocBLAS$ dpkg -l | grep hip

ii  hip_base                   1.5.19284          amd64              HIP: Heterogenous-computing Interface for Portability [BAS

ii  hip_doc                    1.5.19284          amd64              HIP: Heterogenous-computing Interface for Portability [DOC

ii  hip_hcc                    1.5.19284          amd64              HIP: Heterogenous-computing Interface for Portability [HCC

ii  hip_samples                1.5.19284          amd64              HIP: Heterogenous-computing Interface for Portability [SAM

ii  whiptail                   0.52.18-1ubuntu2   amd64              Displays user-friendly dialog boxes from shell scripts

diego@diego-MS-7B09:[master]~/Git/rocBLAS$

--- a/header_compilation_tests.sh
+++ b/header_compilation_tests.sh
@@ -2,15 +2,22 @@
 
 export LOCALE=C
 set -e
-
 exec 2>&1
 
-if [[ ! -e build/release/include/rocblas-export.h ]]; then
+script=$(realpath "$0")
+
+build_first()
+{
     echo "Please run this script after at least one build of rocBLAS."
     exit 1
-fi
+}
 
-script=$(realpath "$0")
+[[ ! -e CMakeCache.txt ]] && build_first
+
+BUILD_DIR=$(realpath "$(pwd)")
+SOURCE_DIR=$(realpath -m "$(grep CMAKE_HOME_DIRECTORY CMakeCache.txt | sed 's/CMAKE_HOME_DIRECTORY:INTERNAL=//g')")
+
+[[ ! -e $BUILD_DIR/include/rocblas-export.h ]] && build_first
 
 # Returns whether the output file is up to date.
 # Prints the output file.
@@ -20,7 +27,7 @@ script=$(realpath "$0")
 out_uptodate()
 {
     local file="$1_$2"
-    local filename="build/compilation_tests/$file.o"
+    local filename="$BUILD_DIR/compilation_tests/$file.o"
     mkdir -p $(dirname "$filename")
     local out=$(realpath -m "$filename")
     echo "$out"
@@ -33,7 +40,7 @@ out_uptodate()
 
 HCC=/opt/rocm/hcc/bin/hcc
 
-HCC_OPTS="-Werror -DBUILD_WITH_TENSILE=1 -DTensile_RUNTIME_LANGUAGE_HIP=1 -DTensile_RUNTIME_LANGUAGE_OCL=0 -Drocblas_EXPORTS -I$(realpath library/include) -I$(realpath library/src/include) -I$(realpath build/release/include) -I$(realpath library/src/blas3/Tensile) -isystem /opt/rocm/hip/include -isystem /opt/rocm/hsa/include -isystem /opt/rocm/hcc/include -isystem /opt/rocm/include -I$(realpath build/release/Tensile) -O3 -DNDEBUG -fPIC -fvisibility=hidden -fvisibility-inlines-hidden -Wno-unused-command-line-argument"
+HCC_OPTS="-Werror -DBUILD_WITH_TENSILE=1 -DTensile_RUNTIME_LANGUAGE_HIP=1 -DTensile_RUNTIME_LANGUAGE_OCL=0 -Drocblas_EXPORTS -I$(realpath library/include) -I$(realpath library/src/include) -I$(realpath $BUILD_DIR/include) -I$(realpath $SOURCE_DIR/library/src/blas3/Tensile) -isystem /opt/rocm/hip/include -isystem /opt/rocm/hsa/include -isystem /opt/rocm/hcc/include -isystem /opt/rocm/include -I$(realpath $BUILD_DIR/Tensile) -O3 -DNDEBUG -fPIC -fvisibility=hidden -fvisibility-inlines-hidden -Wno-unused-command-line-argument"
 
 GPU_OPTS="-hc -fno-gpu-rdc --amdgpu-target=gfx803 --amdgpu-target=gfx900 --amdgpu-target=gfx906 -Werror"
 
@@ -99,12 +106,12 @@ EOF
         exit 1
 fi
 
-# The headers in library/include must compile with clang host, C99 or C++11,
+# The headers in $SOURCE_DIR/library/include must compile with clang host, C99 or C++11,
 # for client code.
 #
 if [[ -x "$CLANG" ]]; then
     xargs_coproc
-    for file in library/include/*.{h,in}; do
+    for file in $SOURCE_DIR/library/include/*.{h,in}; do
         out=$(out_uptodate $file clang) || \
              echo "$CLANG $CLANG_OPTS -c -o "$out" $HCC_OPTS "$file" || (rm -f "$out"; echo "$file" >&4; exit 255)" >&$XARGS_IN
     done
@@ -116,7 +123,7 @@ The public header file $XARGS_OUTPUT cannot be compiled with
 clang host-only compiler. rocBLAS public header files need to be compatible
 with host-only compilers.
 
-<hip/hip_runtime.h> and (sometimes due to bugs) <hip/hip_runtime_api.h> are
+<hip/hip_runtime.h> (and, sometimes due to bugs, <hip/hip_runtime_api.h>) are
 incompatible with C, so they should only be included in the rocBLAS internal
 C++ implemenation, not in the public headers, which must be compatible with C.
 EOF
@@ -125,7 +132,7 @@ EOF
 fi
 
 xargs_coproc
-for file in library/include/*.{h,in}; do
+for file in $SOURCE_DIR/library/include/*.{h,in}; do
     out=$(out_uptodate $file c99) || \
         echo "$C99 -c -o "$out" $HCC_OPTS $GPU_OPTS "$file" || (rm -f "$out"; echo "$file" >&4; exit 255)" >&$XARGS_IN
 done
@@ -144,7 +151,7 @@ EOF
 fi
 
 xargs_coproc
-for file in library/include/*.{h,in}; do
+for file in $SOURCE_DIR/library/include/*.{h,in}; do
     out=$(out_uptodate $file cpp11) ||
         echo "$CPP11 -c -o "$out" $HCC_OPTS $GPU_OPTS "$file" || (rm -f "$out"; echo "$file" >&4; exit 255)" >&$XARGS_IN
 done

--- a/library/src/CMakeLists.txt
+++ b/library/src/CMakeLists.txt
@@ -214,5 +214,5 @@ add_custom_command(
   TARGET rocblas
   POST_BUILD
   COMMAND ${CMAKE_HOME_DIRECTORY}/header_compilation_tests.sh
-  WORKING_DIRECTORY ${CMAKE_HOME_DIRECTORY}
- )
+  WORKING_DIRECTORY ${CMAKE_BINARY_DIR}
+)

